src/iop/basecurve.c

```diff
@@ -302,18 +342,340 @@ int process_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, cl_m
 }
 #endif
 
+static inline void apply_ev_and_curve(
+    const float *const in,
+    float *const out,
+    const int width,
+    const int height,
+    const float mul,
+    const float *const table,
+    const float *const unbounded_coeffs)
+{
+#ifdef _OPENMP
+#pragma omp parallel for default(none) schedule(static)
+#endif
+  for(size_t k = 0; k < (size_t)width * height; k++)
+  {
+    const float *inp = in + 4 * k;
+    float *outp = out + 4 * k;
+    for(int i = 0; i < 3; i++)
+    {
+      const float f = inp[i] * mul;
+      // use base curve for values < 1, else use extrapolation.
+      if(f < 1.0f)
+        outp[i] = table[CLAMP((int)(f * 0x10000ul), 0, 0xffff)];
+      else if(unbounded_coeffs)
+        outp[i] = dt_iop_eval_exp(unbounded_coeffs, f);
+      else outp[i] = 1.0f;
+    }
+  }
+}
+
+static inline void compute_features(
+    float *const col,
+    const int wd,
+    const int ht)
+{
+  // features are product of
+  // 1) well exposedness
+  // 2) saturation
+  // 3) local contrast (handled in laplacian form later)
+#ifdef _OPENMP
+#pragma omp parallel for default(none) schedule(static) collapse(2)
+#endif
+  for(int j=0;j<ht;j++) for(int i=0;i<wd;i++)
+  {
+    const size_t x = 4*(wd*j+i);
+    const float max = MAX(col[x], MAX(col[x+1], col[x+2]));
+    const float min = MIN(col[x], MIN(col[x+1], col[x+2]));
+    const float sat = .1f + .1f*(max-min)/MAX(1e-4, max);
+    col[x+3] = sat;
+
+    float v = fabsf(col[x]-0.54f);
+    v = MAX(fabsf(col[x+1]-0.54f), v);
+    v = MAX(fabsf(col[x+2]-0.54f), v);
+    const float var = 0.5;
+    const float exp = .2f + dt_fast_expf(-v*v/(var*var));
+    col[x+3] *= exp;
+  }
+}
+
+static inline void gauss_blur(
+    const float *const input,
+    float *const output,
+    const int wd,
+    const int ht)
+{
+  const float w[5] = {1./16., 4./16., 6./16., 4./16., 1./16.};
+  float *tmp = dt_alloc_align(64, wd*ht*4*sizeof(float));
```

Comment:
sizeof() is size_t,
but i can not tell whether it will be computed in `max(typeof(wd), typeof(ht), typeof(4), typeof(sizeof))`, or the `typeof(wd)`, and then just cast to the bigger storage size
but if we cast `wd` to size_t first, it won't matter